### DISCUSSION TITLE
Global store support

### DISCUSSION THREAD

**Discussion by pi0** [nuxt/nuxt#16994] [RFCs]
With the introduction of composition-API, it makes less sense of using classic vuex store patterns. In the meantime, having a global state (SSR/SPA) is usually required by the projects and with nuxt, we need to provide a built-in integration layer.

While in vue2 ecosystem, `vuex` was the only option, now we have at least two undergoing projects:

- Pinia (by @posva): https://github.com/posva/pinia  (nuxt module: https://github.com/posva/pinia/tree/v2/packages/nuxt)
- Vuex (https://next.vuex.vuejs.org/) and [Vuex@5 POC](https://github.com/kiaking/vuex-ideas) by @kiaking with [example nuxt 3 plugin](https://github.com/nuxt/framework/blob/c8f86914962bbbca900415aea26de95f71653059/packages/nuxt3/src/app/plugins/vuex.ts)
- ~~Possibility of introducing `useStore` and more serialize/retrieval utilities into vue 3 core (back portable via `@vue/composition-api`?)~~ [useState](https://v3.nuxtjs.org/docs/usage/state)

As currently there are not enough data to pick one approach, it makes sense to not directly choose one library but help to make all different store implementations working nicely with nuxt 3 via each own's module and adding examples, docs for them to the framework. This way, `store/` support can be added with `@pinia/nuxt` or `@nuxtjs/vuex` modules)

When we have a better idea of vue 3's take on the store, we can start drafting built-in store support. Hopefully core, Pinia and Vuex 5 will have almost identical API that we can use it for nuxt 3 layer and allow to switch provider by users.

**Comment by AndreyYolkin** [comment]
So, some modules like @nuxtjs/auth and @nuxtjs/auth-next require vuex store. It causes some difficulties with migration, even in bridge

**Reply by pi0** [reply]
Indeed. For auth, i plan to use [useState](https://v3.nuxtjs.org/docs/usage/state) with nitro storage for universal storage. This we no longer depend on vuex

**Reply by DamianGlowala** [reply]
@pi0 Does it mean there are plans of resuming work on 5th version of `@nuxtjs/auth-next` module (which currently appears to be put on hold)? 🎉

**Reply by pi0** [reply]
Yes! We had no choice but to hold it until we can move minimum dependency to bridge/nitro and leverage new functionalities like capi, useState, and most importantly nitro logic for server auth handling. 🙈

**Reply by thibautSchmutz** [reply]
@pi0 Great idea ! By the way, the doc shows us how to initialize the state but how can we update it ? Also, is the state reactive by default ?

**Reply by danielroe** [reply]
`useState` returns a ref that can be mutated. It is reactive. Example:

```js
const myState = useState('myState', () => ({}))
myState.value.test = true
```

**Reply by wobsoriano** [reply]
> `useState` returns a ref that can be mutated. It is reactive. Example:
> 
> ```js
> const myState = useState('myState', () => ({}))
> myState.value.test = true
> ```

Do we type it like how we type a normal ref?

```ts
const user = useState<User>('user')
```

**Reply by danielroe** [reply]
Yes. 👍

**Reply by midblue** [reply]
Regarding typing, it seems like typed state elements' types sometimes transfer between components/pages in the app and sometimes don't, which makes sense because it's not always clear which will initialize/access it first. Is there any best practice for declaring state element types in a place that will be guaranteed to communicate their types to other elements of the app?

**Comment by mxlpitelik** [comment]
Agreed, al my project use Vuex as business logic layer, it super important, but if you add posibility of different stores adding - it will be super!

Because type difinition in vuex is HELL :(((

P.S. - MobX with proxies superfast now....

**Reply by wobsoriano** [reply]
> er important, but if you add posibility of di

well there's pinia if type defs is ur problem

**Comment by maximilianmikus** [comment]
Having clear goto solutions is a big plus for the vue/nuxt ecosystem. I think its better to support one standard solution everywhere. This creates more synergies with other projects/modules (e.g. with vuex used basically everywhere). It would also be easier to migrate v2 apps hopefully.

**Comment by jmroon** [comment]
Considering Pinia is becoming the official recommendation in the new vue docs, I believe that should be the supported standard at a minimum:
https://staging.vuejs.org/guide/scaling-up/state-management.html#simple-state-management-with-reactivity-api

**Reply by Lycoon** [reply]
Is there any update regarding this for its replacement into Nuxt?

**Reply by woutercouvaras** [reply]
I found this thread asking the similar questions - looks like it's pretty easy to set up and get going:  https://pinia.vuejs.org/ssr/nuxt.html#installation

It might not be the default, but in the world of riding the "bleeding edge" wave of tech, this seems like a low barrier to entry.

**Reply by danielroe** [reply]
This is a PR to follow: https://github.com/vuejs/pinia/pull/791

**Comment by mpietrucha** [comment]
When I started build first app with nuxt3 I discovered that keeping store modules in composables directory will be very messy in the future. So I created very small package that add auto imports from store directory. Works well with pinia.

https://github.com/mpietrucha/nuxt3-store-autoimport

**Comment by richardeschloss** [comment]
FWIW, if anyone is looking for Vuex support, I just released a module: https://github.com/richardeschloss/nuxt3-vuex . This will let you keep using your Nuxt2 "store" folder in Nuxt 3 (with some added features). This may come in handy if you have a massive Nuxt2 project (with a massive store) you wish to migrate to Nuxt 3 :rocket: :rocket: :rocket:

**Reply by MartinX3** [reply]
Thanks, but vuex is deprecated by pinia.
Pinia works just fine with nuxt2/bridge/3.

**Reply by richardeschloss** [reply]
@MartinX3  Pinia is great but there are some breaking changes that wouldn't allow existing Vuex-based store folders in a Nuxt2 project to work. This module was created to support those [very large] legacy projects which may have many files in their store folder which consist of mutations. Upgrading to Pinia would require all existing code that depends on mutations to be updated (i.e., all `$store.commit` calls, potentially 100s of them). I'm not sure how many of those projects want to risk changing their existing global store all at once, even if the change is for the better.

This module would help make a migration easier and risk-free by allowing a Nuxt2's existing store to work, and at the same time, by referencing this discussion thread (in the project's readme), suggest how to *incrementally* upgrade their state management out of Vuex.

**Reply by vedmant** [reply]
@richardeschloss Hey, did you delete your repository?

**Reply by vedmant** [reply]
If someone needs this, I restored this repo https://github.com/vedmant/nuxt3-vuex also added SSR support and nested stores support

**Reply by richardeschloss** [reply]
> @richardeschloss Hey, did you delete your repository?

Github: yes, npm: no (with a deprecation note).
